arangod/Replication2/ReplicatedLog/Components/LogFollower2.cpp

```diff
@@ -0,0 +1,245 @@
+      compaction(std::make_shared<CompactionManager>(*storage, options,
+                                                     loggerContext)),
+      stateHandle(
+          std::make_shared<StateHandleManager>(std::move(stateHandlePtr))),
+      snapshot(std::make_shared<SnapshotManager>(
+          *storage, *stateHandle, termInfo, leaderComm, loggerContext)),
+      commit(std::make_shared<FollowerCommitManager>(*storage, *stateHandle,
+                                                     loggerContext)),
+      appendEntriesManager(std::make_shared<AppendEntriesManager>(
+          termInfo, *storage, *snapshot, *compaction, *commit, metrics,
+          loggerContext)),
+      termInfo(termInfo) {
+  metrics->replicatedLogFollowerNumber->operator++();
+  auto provider = std::make_unique<MethodsProvider>(*this);
+  stateHandle->becomeFollower(std::move(provider));
+}
+
+FollowerManager::~FollowerManager() {
+  metrics->replicatedLogFollowerNumber->operator--();
+}
+
+auto FollowerManager::getStatus() const -> LogStatus {
+  auto commitIndex = commit->getCommitIndex();
+  auto log = storage->getCommittedLog();
+  auto [releaseIndex, lowestIndexToKeep] = compaction->getIndexes();
+  return LogStatus{FollowerStatus{
+      .local =
+          LogStatistics{
+              .spearHead = log.getLastTermIndexPair(),
+              .commitIndex = commitIndex,
+              .firstIndex = log.getFirstIndex(),
+              .releaseIndex = releaseIndex,
+          },
+      .leader = termInfo->leader,
+      .term = termInfo->term,
+      .lowestIndexToKeep = lowestIndexToKeep,
+      .compactionStatus = compaction->getCompactionStatus(),
+      .snapshotAvailable =
+          snapshot->checkSnapshotState() == SnapshotState::AVAILABLE,
+  }};
+}
+
+auto FollowerManager::getQuickStatus() const -> QuickLogStatus {
+  auto commitIndex = commit->getCommitIndex();
+  auto log = storage->getCommittedLog();
+  auto [releaseIndex, largestIndexToKeep] = compaction->getIndexes();
+  return QuickLogStatus{
+      .role = ParticipantRole::kFollower,
+      .localState = stateHandle->getQuickStatus(),
+      .term = termInfo->term,
+      .local =
+          LogStatistics{
+              .spearHead = log.getLastTermIndexPair(),
+              .commitIndex = commitIndex,
+              .firstIndex = log.getFirstIndex(),
+              .releaseIndex = releaseIndex,
+          },
+      .leadershipEstablished = commitIndex.value > 0,
+      .snapshotAvailable =
+          snapshot->checkSnapshotState() == SnapshotState::AVAILABLE,
+  };
+}
+
+auto FollowerManager::resign()
+    -> std::tuple<std::unique_ptr<replicated_state::IStorageEngineMethods>,
+                  std::unique_ptr<IReplicatedStateHandle>, DeferredAction> {
+  // 1. resign the state and receive its handle
+  auto handle = stateHandle->resign();
+  // 2. resign the storage manager to receive the storage engine methods
+  auto methods = storage->resign();
+  // TODO resign CommitManager
```

Comment:
This sounds like it should be resolved, or is there no race that would allow the commit manager to access the storage- or state handle manager after they're destroyed?

Reply:
By design, if you depend on another component, it will outlive you. However, it might refuse to act and return an error instead.

I initially planned to resign the commit manager, but it turns out, that there is nothing to resign, except maybe for some promises.